src/connector/src/pulsar/source/reader.rs

```diff
@@ -13,63 +13,171 @@
 // limitations under the License.
 
 use std::borrow::BorrowMut;
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
 
 use anyhow::{anyhow, Result};
 use async_trait::async_trait;
-use futures::StreamExt;
-use pulsar::{Consumer, Pulsar, TokioExecutor};
+use futures::future::try_join_all;
+use futures::{StreamExt, TryStreamExt};
+use itertools::Itertools;
+use pulsar::consumer::{InitialPosition, Message};
+use pulsar::message::proto::MessageIdData;
+use pulsar::{Consumer, ConsumerBuilder, ConsumerOptions, Pulsar, SubType, TokioExecutor};
+use risingwave_common::try_match_expand;
+use tokio::sync::{mpsc, oneshot};
+use tokio_stream::wrappers::UnboundedReceiverStream;
 
 use crate::base::{SourceMessage, SplitReader};
-use crate::pulsar::split::{PulsarOffset, PulsarSplit};
-use crate::{ConnectorStateV2, Properties};
+use crate::pulsar::split::PulsarSplit;
+use crate::pulsar::{PulsarEnumeratorOffset, PULSAR_CONFIG_SERVICE_URL_KEY};
+use crate::{AnyhowProperties, ConnectorStateV2, Properties, SplitImpl};
 
-pub struct PulsarSplitReader {
+struct PulsarSingleSplitReader {
     pulsar: Pulsar<TokioExecutor>,
     consumer: Consumer<Vec<u8>, TokioExecutor>,
     split: PulsarSplit,
 }
 
+fn parse_message_id(id: &str) -> Result<MessageIdData> {
+    let splits = id.split(':').collect_vec();
+
+    if splits.len() < 2 || splits.len() > 4 {
+        return Err(anyhow!("illegal message id string {}", id));
+    }
+
+    let ledger_id = splits[0]
+        .parse::<u64>()
+        .map_err(|e| anyhow!("illegal ledger id {}", e))?;
+    let entry_id = splits[1]
+        .parse::<u64>()
+        .map_err(|e| anyhow!("illegal entry id {}", e))?;
+
+    let mut message_id = MessageIdData {
+        ledger_id,
+        entry_id,
+        partition: None,
+        batch_index: None,
+        ack_set: vec![],
+        batch_size: None,
+    };
+
+    if splits.len() > 2 {
+        let partition = splits[2]
```

Comment:
I don't get why need treats `splits` here. & maybe use a match here

```rust
match splits.len() {
    2..4 => {},
}
```

Reply:
Because the offset of pulsar is represented by a string of four numbers, but the last two can have default values, so they need to be handled